test/topology_custom/test_raft_recovery_user_data.py

```diff
@@ -0,0 +1,192 @@
+    continue succeeding since at least two nodes are alive at any point during the recovery procedure (it involves only
+    rolling restarts), every node in dc1 is a replica (3 nodes, RF=3) and two nodes make a local quorum in dc1.
+    5. Remove nodes from dc2 from topology using remove or replace, depending on the value of the remove_dead_nodes_with
+    parameter. For remove, we must do two additional steps to make it work:
+    - Mark all dead nodes as permanently dead.
+    - Decrease RF of the user keyspace, which uses tablets, to 0 in dc2.
+    6. For remove, add 3 nodes to dc2 and bring back RF of the user keyspace to 3 in dc2, which brings back dc2 to its
+    previous state. For replace, add a single node (a sanity check verifying that the cluster is functioning properly).
+    7. Stop sending writes.
+    """
+    # Decrease failure_detector_timeout_in_ms from the default 20 s to speed up some graceful shutdowns in the test.
+    # Shutting down the CQL server can hang for failure_detector_timeout_in_ms in the presence of dead nodes and
+    # CQL requests.
+    cfg = {
+        'endpoint_snitch': 'GossipingPropertyFileSnitch',
+        'enable_tablets': True,
+        'failure_detector_timeout_in_ms': 2000,
+    }
+    property_file_dc1 = {'dc': 'dc1', 'rack': 'rack1'}
+    property_file_dc2 = {'dc': 'dc2', 'rack': 'rack2'}
+
+    logging.info('Adding servers that will survive majority loss to dc1')
+    live_servers = await manager.servers_add(3, config=cfg, property_file=property_file_dc1)
+    logging.info('Adding servers that will be killed to dc2')
+    dead_servers = await manager.servers_add(3, config=cfg, property_file=property_file_dc2)
+    logging.info(f'Servers to survive majority loss: {live_servers}, servers to be killed: {dead_servers}')
+
+    cql, _ = await manager.get_ready_cql(live_servers + dead_servers)
+    hosts = await wait_for_cql_and_get_hosts(cql, live_servers, time.time() + 60)
+
+    first_group0_id = (await cql.run_async(
+            "SELECT value FROM system.scylla_local WHERE key = 'raft_group0_id'"))[0].value
+
+    rf: int = 3
+    ks_name = unique_name()
+    finish_writes = await start_writes(cql, rf, ConsistencyLevel.LOCAL_QUORUM, 5, ks_name, True)
+
+    # Send some writes before we kill nodes.
+    logging.info('Sleeping for 1 s')
+    await asyncio.sleep(1)
+
+    logging.info(f'Killing {dead_servers}')
+    for srv in dead_servers:
+        await manager.server_stop(server_id=srv.server_id)
+
+    logging.info('Starting the recovery procedure')
+
+    logging.info(f'Restarting {live_servers}')
+    await manager.rolling_restart(live_servers)
+
+    logging.info(f'Deleting the persistent discovery state and group 0 ID on {live_servers}')
+    for h in hosts:
+        await delete_discovery_state_and_group0_id(cql, h)
+
+    recovery_leader_id = await manager.get_host_id(live_servers[0].server_id)
```

Comment:
We should change this to use the leader-finding API that you plan to implement in a follow-up. I'm not sure we have a guarantee here that the 0-th server will have the freshest state; in theory, a background operation might change the group0 state without your knowledge while you are killing the majority of the cluster, so I'm not sure that at this point every node in the cluster - or, at least node 0 - will have the newest state.

Reply:
You are right, a background operation could change the group 0 state just before losing majority. It's not only in theory, but it can really happen in this test. We are sending writes while losing majority, so a tablet migration can start at any moment. I even had to disable tablet migrations below to unflake the`check_system_topology_and_cdc_generations_v3_consistency` calls in this test.

I'll add a FIXME to use the leader-finding API.

Reply:
Done, I also updated to comment with follow-ups to not forget about it.